Security Groups for ELBs

mattmb opened the pull request:
This adds the possibility to add security groups for your ELBs, it also defines a default ELB SG of open to the world on 80/443.

Commit:
Security Groups for ELBs

This adds the possibility to add security groups for your ELBs,
it also defines a default ELB SG of open to the world on 80/443.

ashb commented on tests/tests.py:
```diff
@@ -257,7 +260,7 @@ def test_elb(self):
                 'tests/sample-project.yaml',
                 'dev').config,'my-stack-name')
-        self.assertEquals(known, config.elb())
+        compare(known, config.elb()[0])
```
This looks like a suspect – can you assidn `elb, elb_sg` and then compare elb directly as it's clearer what this is doing.

mattmb replied:
Good shout

ashb commented on tests/tests.py:
```diff
@@ -365,7 +369,7 @@ def test_elb_with_ssl(self):
         }]
         config = ConfigParser(project_config.config,'my-stack-name')
-        self.assertEquals(known, config.elb())
+        self.assertEquals(known, config.elb()[0])
```
Same here

ashb commented on tests/tests.py:
```diff
@@ -413,7 +418,7 @@ def test_elb_with_reserved_chars(self):
         }]
         config = ConfigParser(project_config.config,'my-stack-name')
-        self.assertEquals(known, config.elb())
+        self.assertEquals(known, config.elb()[0])
```
And one more.

ashb commented:
It looks like this will have a default SG on the ELB allowing 80 and 443 from everyone, but if you specify any SGs for your ELB it will be replaced and the default one won't be used anymore?

Is that right? If so can you make that clearer in the readme please?

mattmb commented:
Yep, you got it. Will add a comment to make it clear.

mattmb commented:
@ashb you're right this does allow us to give only ELBs access to the hosts, solving #13 .

See https://github.com/ministryofjustice/template-deploy/pull/5 for how.

ashb commented on bootstrap_cfn/config.py:
```diff
@@ -225,6 +227,17 @@ def elb(self):
+                for sg_name, sg in elb['security_groups'].items():
+                    new_sg = deepcopy(elb_sg)
+                    new_sg['Properties']['SecurityGroupIngress'] = sg
+                    elb_sgs[sg_name + safe_name] = new_sg
```
Security group names can contain "a-z, A-Z, 0-9, spaces, and ._-:/()#,@[]+=&;{}!$*" so we should be able to use the raw SG name without modification - this makes it clearer to parse the SG name - currently we would end up with something like "DefaultSGtesttemplateext" but "DefaultSG-test-template-ext" would be much preferable.

ashb replied:
The above commend applies more to line 238 `['DefaultSG' + safe_name]`.

For this one: it's slightly unexpected that the SG name you put in the yaml isn't the SG name you'd put elsewhere to reference it...

Commit:
SG Names for ELBs

For names specified in the config, there is no need to append the
ELB name. This makes it easier to refer to these SGs elsewhere in
the config. For default SGs we must still append the ELB name to
ensure it is unique in the CF json.
